azure_functions_worker/dispatcher.py

```diff
@@ -313,82 +322,92 @@ async def _handle__worker_status_request(self, request):
             request_id=request.request_id,
             worker_status_response=protos.WorkerStatusResponse())
 
-    async def _handle__functions_metadata_request(self, request):
-        metadata_request = request.functions_metadata_request
-        directory = metadata_request.function_app_directory
+    def get_function_metadata(self, directory, caller_info):
+        """
+        This method is called to index the functions in the function app
+        directory and save the results in function_metadata_result or
+        function_metadata_exception in case of an exception.
+        """
         script_file_name = get_app_setting(
             setting=PYTHON_SCRIPT_FILE_NAME,
             default_value=f'{PYTHON_SCRIPT_FILE_NAME_DEFAULT}')
-        function_path = os.path.join(directory, script_file_name)
 
         logger.info(
-            'Received WorkerMetadataRequest, request ID %s, function_path: %s',
-            self.request_id, function_path)
+            'Received WorkerMetadataRequest from %s, request ID %s, '
+            'script_file_name: %s',
+            caller_info, self.request_id, script_file_name)
 
         try:
             validate_script_file_name(script_file_name)
+            function_path = os.path.join(directory, script_file_name)
 
-            if not os.path.exists(function_path):
-                # Fallback to legacy model
-                return protos.StreamingMessage(
-                    request_id=request.request_id,
-                    function_metadata_response=protos.FunctionMetadataResponse(
-                        use_default_metadata_indexing=True,
-                        result=protos.StatusResult(
-                            status=protos.StatusResult.Success)))
+            self.function_metadata_result = (
+                self.index_functions(function_path)) \
+                if os.path.exists(function_path) else None
 
-            fx_metadata_results = self.index_functions(function_path)
+        except Exception as ex:
+            self.function_metadata_exception = self._serialize_exception(ex)
```

Comment:
if worker failed after exception is assigned and before handle_metadata_request is invoked. this error is swallowed. i think we shall raise this in init response, let me know how u think

Reply:
same case for handle_env_reload

Reply:
This exception should be saved and raised only in the metadata response since it is related to indexing. We should not throw these exceptions in init or reload since host treats these requests to complete a specific task.
Saving the error and throwing it in the metadata response will also help us debug at which stage the function app failed with the correct error message.

Reply:
if it is second worker or later, worker wont receive metadatarequest right? i think we need to at least log the exception when doing the indexing to ensure the exception always gets captured in the log. ex. env_reload succeeded but func indexing actually failed, if for some reason handle metadatarequest is not called, the exception log still shall be recorded for troubleshooting.
![image](https://github.com/Azure/azure-functions-python-worker/assets/12449837/4dbdd4df-50f7-4865-b18d-2543b0a279eb)

Reply:
Good point. Fixed this.